sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java

```diff
@@ -0,0 +1,276 @@
+
+    }
+
+    @Override
+    public EvaluationOutcome evaluate(MesosResourcePool mesosResourcePool, PodInfoBuilder podInfoBuilder) {
+
+        // Generate new private key and encode it to PEM
+        // Generate new CSR, sign it, encode certificate to PEM
+        try {
+            KeyPair keyPair = keyPairGenerator.generateKeyPair();
+
+            X509Certificate certificate = certificateAuthorityClient.sign(generateCSR(keyPair));
+
+            String certPEM = certificateToPem(certificate);
+            String privateKeyPEM = privateKeyToPem(keyPair.getPrivate());
+
+            storeSecrets(podInfoBuilder, certPEM, privateKeyPEM);
+        } catch (Exception e) {
+            LOGGER.error("Failed to get certificate", e);
+            return EvaluationOutcome.fail(
+                    this, "Failed because of exception: %s", e);
+        }
+
+        Collection<Protos.Volume> volumes = getExecutorInfoSecretVolumes(podInfoBuilder);
+
+        // Share keys to the container
+        Optional<Protos.ExecutorInfo.Builder> executorBuilder = podInfoBuilder.getExecutorBuilder();
+        Protos.TaskInfo.Builder taskBuilder = podInfoBuilder.getTaskBuilder(taskName);
+        if (executorBuilder.isPresent()) {
+            executorBuilder.get()
+                    .getContainerBuilder()
+                    .setType(Protos.ContainerInfo.Type.MESOS)
+                    .addAllVolumes(volumes);
+            taskBuilder.setExecutor(executorBuilder.get());
+        }
+
+        return EvaluationOutcome.pass(
+                this, null, "TLS certificate created and added to the pod");
+
+    }
+
+    private Collection<Protos.Volume> getExecutorInfoSecretVolumes(PodInfoBuilder podInfoBuilder) {
+        String filenameInContainer = podInfoBuilder
+                .getPodInstance()
+                .getPod()
+                .getTransportEncryption()
+                .get()
+                .getName();
+
+        HashMap<String, String> tlsSecrets = new HashMap<>();
+        tlsSecrets.put(
+                getSecretPath(podInfoBuilder, "certificate-crt"),
+                String.format("%s.crt", filenameInContainer));
+
+        tlsSecrets.put(
+                getSecretPath(podInfoBuilder, "private-key"),
+                String.format("%s.key", filenameInContainer));
+
+        Collection<Protos.Volume> volumes = new ArrayList<>();
+
+        for (Map.Entry<String, String> entry : tlsSecrets.entrySet()) {
+
+            volumes.add(Protos.Volume.newBuilder()
+                    .setSource(Protos.Volume.Source.newBuilder()
+                            .setType(Protos.Volume.Source.Type.SECRET)
+                            .setSecret(getReferenceSecret(entry.getKey()))
+                            .build())
+                    .setContainerPath(entry.getValue())
+                    .setMode(Protos.Volume.Mode.RO)
+                    .build());
+
+        }
```

Comment:
Can't recall now if you're using Java 8 elsewhere, but this could be replaced with a lambda:
```java
        tlsSecrets.entrySet().forEach(tlsSecretEntry ->
                volumes.add(Protos.Volume.newBuilder()
                        .setSource(Protos.Volume.Source.newBuilder()
                                .setType(Protos.Volume.Source.Type.SECRET)
                                .setSecret(getReferenceSecret(tlsSecretEntry.getKey()))
                                .build())
                        .setContainerPath(tlsSecretEntry.getValue())
                        .setMode(Protos.Volume.Mode.RO)
                        .build())
        );
```